src/experiments.js

```diff
@@ -88,15 +92,64 @@ class Test extends HTMLElement {
 		} ) );
 	}
 
+	getVariantId() {
+		const testId = this.testIdWithPost;
+		const trafficPercentage = this.trafficPercentage;
+
+		// Check if this user already have a variant for this test.
+		const currentTests = this.getTestsForUser();
+		let variantId = false;
+		// Test variant can be 0 so check for not undefined and not strictly false and
+		// that it's a valid index.
+		if (
+			typeof currentTests[ testId ] !== 'undefined' &&
+			currentTests[ testId ] !== false &&
+			currentTests[ testId ] < this.variants.length
+		) {
+			variantId = currentTests[ testId ];
+		} else if ( currentTests[ testId ] === false ) {
+			return variantId;
+		} else {
```

Comment:
Curious to see if we can shorten this is a bit, just to see if there is a case I couldn't account for:
```suggestion
		let variantId = currentTests[ testId ] || false;
		if ( ! variantId ) {
```

since :
```js
const values = ['x'];
// All the below are true
console.log( values[undefined] === undefined );
console.log( values['string'] === undefined );
console.log( values[false] === undefined );
console.log( values[10] === undefined );
```

Reply:
I think this does do the job but it's losing 2 things:

- basic `currentTests[ testId ]` value validation from `currentTests[ testId ] < this.variants.length`
- early return if `currentTests[ testId ]` is set but the value is `false`

We need to differentiate between undefined and `false` in terms of the value - `false` means the visitor is excluded from the test, while `undefined` means determine whether they should be in the test, then if so which variant they should get.

Maybe something like:

```suggestion
		let variantId = false;
		// Test variant can be 0 so check for not undefined and not strictly false and
		// that it's a valid index.
		let variantId = currentTests[ testId ];
		if ( variantId === false ) {
			return variantId;
		} else if ( typeof variantId === 'undefined' || variantId >= this.variants.length ) {
```

Reply:
I'll leave it this way for now but I'll add an issue for a refactor of this file. It could do with splitting out and consolidating further really.

Reply:
Sure, it's not a blocker. And I didn't factor the length case. 👍